// admin-on-rest/lib/auth/Firewall.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _reactRouterRedux = require('react-router-redux');

var _ = require('./');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Restricted = function (_Component) {
    (0, _inherits3.default)(Restricted, _Component);

    function Restricted() {
        var _ref;

        var _temp, _this, _ret;

        (0, _classCallCheck3.default)(this, Restricted);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Restricted.__proto__ || Object.getPrototypeOf(Restricted)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
            isAuthenticated: false
        }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
    }

    (0, _createClass3.default)(Restricted, [{
        key: 'componentWillMount',
        value: function componentWillMount() {
            var _this2 = this;

            var _props = this.props,
                authClient = _props.authClient,
                authParams = _props.authParams,
                location = _props.location,
                replace = _props.replace;

            authClient(_.AUTH_CHECK, authParams).then(function () {
                return _this2.setState({ isAuthenticated: true });
            }).catch(function (e) {
                return replace({
                    pathname: e && e.redirectTo || '/login',
                    state: { nextPathname: location.pathname }
                });
            });
        }
    }, {
        key: 'render',
        value: function render() {
            var isAuthenticated = this.state.isAuthenticated;
            var _props2 = this.props,
                children = _props2.children,
                rest = (0, _objectWithoutProperties3.default)(_props2, ['children']);

            return isAuthenticated ? _react2.default.cloneElement(children, rest) : null;
        }
    }]);
    return Restricted;
}(_react.Component);

Restricted.propTypes = {
    authClient: _react.PropTypes.func,
    authParams: _react.PropTypes.object,
    location: _react.PropTypes.object,
    replace: _react.PropTypes.func
};
exports.default = (0, _reactRedux.connect)(null, {
    replace: _reactRouterRedux.replace
})(Restricted);
module.exports = exports['default'];
